source/opt/folding_rules.cpp

```diff
@@ -2296,6 +2297,48 @@ FoldingRule BitCastScalarOrVector() {
   };
 }
 
+// Remove indirect bitcasts which have no effect.
+//   uint32 x;  asuint32(x)            => x
+//   uint32 x;  asuint32(asint32(x))   => x
+//   float32 x; asuint32(asint32(x))   => asuint32(x)
+FoldingRule RedundantBitcast() {
+  return [](IRContext* context, Instruction* inst,
+            const std::vector<const analysis::Constant*>&) {
+    assert(inst->opcode() == spv::Op::OpBitcast);
+
+    analysis::DefUseManager* def_mgr = context->get_def_use_mgr();
+    Instruction* child = def_mgr->GetDef(inst->GetSingleWordInOperand(0));
+
+    if (inst->type_id() == child->type_id()) {
+      inst->SetOpcode(spv::Op::OpCopyObject);
+      inst->SetInOperands({{SPV_OPERAND_TYPE_ID, {child->result_id()}}});
+      return true;
+    }
+
+    if (child->opcode() != spv::Op::OpBitcast) {
+      return false;
+    }
+    analysis::TypeManager* type_mgr = context->get_type_mgr();
+    const analysis::Type* inst_type = type_mgr->GetType(inst->type_id());
+    const analysis::Type* child_type = type_mgr->GetType(child->type_id());
+
+    if (inst_type->NumberOfComponents() != child_type->NumberOfComponents()) {
+      return false;
+    }
+    if (ElementWidth(inst_type) != ElementWidth(child_type)) {
+      return false;
+    }
```

Comment:
What case is this trying to avoid? Why can't a cast from 4xi32 to 2xi64 and then back to 4xi32 be folded? I know the wording in the spec is hard to parse, but this should be allowed.

Reply:
In this case that is try 4 is an integer multiple of 2. It is oddly worded, but the meaning is that the first two elements of the 4xi32 map to the first element of the 2xi64. When you go to convert it back, the mapping should be the same, so it can be reversed. You can remove those checks.